File: packages/pip/src/goodvibes_cli/utils/json_merge.py
"""Merge goodvibes-managed keys into user-modified settings.json / .mcp.json."""
from __future__ import annotations

import copy
import json
import os
import pathlib
import re
import shutil

MANAGED_JSON = [".claude/settings.json", ".mcp.json"]

_MARKER = re.compile(r"^: (goodvibes-[a-z0-9-]+);")


def _marked_hook(group: dict) -> tuple[int, str] | None:
    hooks = group.get("hooks") if isinstance(group, dict) else None
    for i, h in enumerate(hooks if isinstance(hooks, list) else []):
        cmd = h.get("command") if isinstance(h, dict) else None
        m = _MARKER.match(cmd) if isinstance(cmd, str) else None
        if m:
            return i, m.group(1)
    return None


def _hook_id(group: dict) -> str | None:
    found = _marked_hook(group)
    return found[1] if found else None


def write_json(path: pathlib.Path, data: object) -> None:
    """Write via a temp file in the same folder so a crash never leaves a half-written settings file."""
    tmp = path.with_name(f".{path.name}.{os.getpid()}.tmp")
    try:
        with open(tmp, "x", encoding="utf-8", newline="\n") as f:
            f.write(json.dumps(data, indent=2, ensure_ascii=False) + "\n")
        if path.exists():
            shutil.copymode(path, tmp)
        os.replace(tmp, path)
    except BaseException:
        tmp.unlink(missing_ok=True)
        raise


def managed_ids(rel: str, tpl: dict) -> list[str]:
    if rel == ".mcp.json":
        return [f"mcp:{name}" for name in (tpl.get("mcpServers") or {})]
    ids = []
    for lst in ("ask", "deny"):
        ids += [f"{lst}:{p}" for p in (tpl.get("permissions") or {}).get(lst) or []]
    for event, groups in (tpl.get("hooks") or {}).items():
        for g in groups:
            hid = _hook_id(g)
            if hid:
                ids.append(f"hook:{event}:{hid}")
    return ids


def present_ids(rel: str, tpl: dict, content: dict) -> list[str]:
    if not isinstance(content, dict):
        return []

    def present(mid: str) -> bool:
        kind, rest = mid.split(":", 1)
        if kind == "mcp":
            return rest in (content.get("mcpServers") or {})
        if kind == "hook":
            event, hid = rest.split(":", 1)
            return any(_hook_id(g) == hid for g in (content.get("hooks") or {}).get(event) or [])
        return rest in ((content.get("permissions") or {}).get(kind) or [])

    return [mid for mid in managed_ids(rel, tpl) if present(mid)]


# Allow rules goodvibes shipped up to 1.9.1: they auto-approved running arbitrary code, so update takes them back out.
RETIRED_ALLOW = ['Bash(npm install*)', 'Bash(npm run*)', 'Bash(npx*)', 'Bash(pip install*)', 'Bash(uv*)', 'Bash(python*)', 'Bash(node*)', 'Bash(git restore *)']


def merge_managed_json(
    rel: str, tpl: dict, user: dict, installed: list[str] | None = None, retire_allow: bool = False
) -> tuple[dict, list[str]]:
    """An id in `installed` but absent from `user` was removed by the user and stays removed."""
    merged = copy.deepcopy(user)
    changes: list[str] = []
    installed = installed or []

    if rel == ".mcp.json":
        for name, server in (tpl.get("mcpServers") or {}).items():
            current = (merged.get("mcpServers") or {}).get(name)
            if current:
                nxt = {**current, **server}
                if nxt != current:
                    merged["mcpServers"][name] = nxt
                    changes.append(f"~ mcpServers.{name}")
            elif f"mcp:{name}" not in installed:
                merged.setdefault("mcpServers", {})[name] = server
                changes.append(f"+ mcpServers.{name}")
        return merged, changes

    allow = (merged.get("permissions") or {}).get("allow")
    if retire_allow and isinstance(allow, list):
        changes.extend(f"- permissions.allow: {p}" for p in allow if p in RETIRED_ALLOW)
        merged["permissions"]["allow"] = [p for p in allow if p not in RETIRED_ALLOW]

    for lst in ("ask", "deny"):
        for p in (tpl.get("permissions") or {}).get(lst) or []:
            have = (merged.get("permissions") or {}).get(lst) or []
            if p in have or f"{lst}:{p}" in installed:
                continue
            merged.setdefault("permissions", {})[lst] = [*have, p]
            changes.append(f"+ permissions.{lst}: {p}")

    for event, groups in (tpl.get("hooks") or {}).items():
        for g in groups:
            hid = _hook_id(g)
            if not hid:
                continue
            user_groups = (merged.get("hooks") or {}).get(event) or []
            idx = next((i for i, ug in enumerate(user_groups) if _hook_id(ug) == hid), -1)
            if idx >= 0:
                # Only the marked hook is goodvibes'; the user's other hooks and fields in that group stay.
                user_hooks = user_groups[idx]["hooks"]
                j = _marked_hook(user_groups[idx])[0]
                tpl_hook = g["hooks"][_marked_hook(g)[0]]
                if user_hooks[j] != tpl_hook:
                    user_hooks[j] = copy.deepcopy(tpl_hook)
                    changes.append(f"~ hooks.{event}: {hid}")
            elif f"hook:{event}:{hid}" not in installed:
                merged.setdefault("hooks", {})[event] = [*user_groups, g]
                changes.append(f"+ hooks.{event}: {hid}")
    return merged, changes


def managed_record(cwd: pathlib.Path, template_dir: pathlib.Path, prev: dict | None = None) -> dict[str, list[str]]:
    """Keeps previously installed ids so a user's deliberate removal survives later updates."""
    prev = prev or {}
    record = dict(prev)
    for rel in MANAGED_JSON:
        tpl_path, dest_path = template_dir / rel, cwd / rel
        if not tpl_path.exists() or not dest_path.exists():
            continue
        try:
            content = json.loads(dest_path.read_text(encoding="utf-8"))
        except ValueError:
            continue  # unparseable user file: keep the previous record rather than guess
        tpl = json.loads(tpl_path.read_text(encoding="utf-8"))
        record[rel] = list(dict.fromkeys([*prev.get(rel, []), *present_ids(rel, tpl, content)]))
    return record

File: packages/npm/src/utils/json-merge.ts
import { readFile } from 'node:fs/promises'
import { existsSync } from 'node:fs'
import { join } from 'node:path'

// MCP files and the key each tool keeps its servers under.
const MCP_KEY: Record<string, string> = { '.mcp.json': 'mcpServers', '.cursor/mcp.json': 'mcpServers', '.vscode/mcp.json': 'servers' }

// Gemini CLI and Codex keep hooks in Claude Code's shape, so their files merge like settings.json.
export const MANAGED_JSON = ['.claude/settings.json', '.gemini/settings.json', '.codex/hooks.json', ...Object.keys(MCP_KEY)]

type Json = Record<string, any>

const MARKER = /^: (goodvibes-[a-z0-9-]+);/

const markerOf = (h: Json): string | null => (typeof h?.command === 'string' ? h.command.match(MARKER)?.[1] ?? null : null)

function hookId(group: Json): string | null {
  for (const h of group?.hooks ?? []) {
    const id = markerOf(h)
    if (id) return id
  }
  return null
}

export const isJsonObject = (v: unknown): v is Json => typeof v === 'object' && v !== null && !Array.isArray(v)

const same = (a: unknown, b: unknown): boolean => JSON.stringify(a) === JSON.stringify(b)

// The merge reads and extends these containers, so a wrong type would crash it or corrupt the file.
export function shapeError(rel: string, content: Json): string | null {
  const obj = (v: unknown, path: string) => (v == null || isJsonObject(v) ? null : `"${path}" is not a JSON object`)
  const arr = (v: unknown, path: string) => (v == null || Array.isArray(v) ? null : `"${path}" is not a JSON array`)
  const key = MCP_KEY[rel]
  if (key) {
    const servers = content[key]
    return obj(servers, key) ?? Object.entries<unknown>(servers ?? {}).map(([n, v]) => obj(v, `${key}.${n}`)).find(Boolean) ?? null
  }
  const problems = [obj(content.permissions, 'permissions'), obj(content.hooks, 'hooks')]
  for (const list of ['allow', 'ask', 'deny']) problems.push(arr(content.permissions?.[list], `permissions.${list}`))
  for (const [event, groups] of Object.entries<unknown>(isJsonObject(content.hooks) ? content.hooks : {})) {
    problems.push(arr(groups, `hooks.${event}`))
    if (Array.isArray(groups)) groups.forEach((g, i) => problems.push(obj(g, `hooks.${event}[${i}]`) ?? arr(g?.hooks, `hooks.${event}[${i}].hooks`)))
  }
  return problems.find(Boolean) ?? null
}

// Ids of every goodvibes-managed key the template defines for this file.
export function managedIds(rel: string, tpl: Json): string[] {
  const ids: string[] = []
  const key = MCP_KEY[rel]
  if (key) {
    for (const name of Object.keys(tpl[key] ?? {})) ids.push(`mcp:${name}`)
    return ids
  }
  for (const list of ['ask', 'deny']) {
    for (const p of tpl.permissions?.[list] ?? []) ids.push(`${list}:${p}`)
  }
  for (const [event, groups] of Object.entries<Json[]>(tpl.hooks ?? {})) {
    for (const g of groups) {
      const id = hookId(g)
      if (id) ids.push(`hook:${event}:${id}`)
    }
  }
  return ids
}

// Managed ids currently present in a file's content.
export function presentIds(rel: string, tpl: Json, content: Json): string[] {
  const splitFirst = (s: string): [string, string] => {
    const i = s.indexOf(':')
    return [s.slice(0, i), s.slice(i + 1)]
  }
  return managedIds(rel, tpl).filter(id => {
    const [kind, rest] = splitFirst(id)
    if (kind === 'mcp') return rest in (content[MCP_KEY[rel]] ?? {})
    if (kind === 'hook') {
      const [event, hid] = splitFirst(rest)
      return (content.hooks?.[event] ?? []).some((g: Json) => hookId(g) === hid)
    }
    return (content.permissions?.[kind] ?? []).includes(rest)
  })
}

// An id in `installed` but absent from `user` was removed by the user and stays removed.
// Allow rules goodvibes shipped earlier: up to 1.9.1 they auto-approved running arbitrary code; Write(**) did nothing and made Claude Code warn.
// Deny rules goodvibes shipped up to 1.10.0: they prefix-match --force-with-lease, so its ask rule never applied.
export const RETIRED_DENY = ['Bash(git push --force*)', 'Bash(git push * --force*)']

export const RETIRED_ALLOW = ['Bash(npm install*)', 'Bash(npm run*)', 'Bash(npx*)', 'Bash(pip install*)', 'Bash(uv*)', 'Bash(python*)', 'Bash(node*)', 'Bash(git restore *)', 'Write(**)']

export function mergeManagedJson(
  rel: string,
  tpl: Json,
  user: Json,
  installed: string[] = [],
  retireAllow = false,
): { merged: Json; changes: string[] } {
  const merged: Json = structuredClone(user)
  const changes: string[] = []
  const wasInstalled = (id: string) => installed.includes(id)

  const key = MCP_KEY[rel]
  if (key) {
    for (const [name, server] of Object.entries<Json>(tpl[key] ?? {})) {
      const current = merged[key]?.[name]
      if (current) {
        const next = { ...current, ...server }
        if (!same(current, next)) {
          merged[key][name] = next
          changes.push(`~ ${key}.${name}`)
        }
      } else if (!wasInstalled(`mcp:${name}`)) {
        merged[key] = { ...(merged[key] ?? {}), [name]: server }
        changes.push(`+ ${key}.${name}`)
      }
    }
    return { merged, changes }
  }

  if (retireAllow && Array.isArray(merged.permissions?.allow)) {
    const keep = merged.permissions.allow.filter((p: string) => !RETIRED_ALLOW.includes(p))
    for (const p of merged.permissions.allow) if (RETIRED_ALLOW.includes(p)) changes.push(`- permissions.allow: ${p}`)
    merged.permissions.allow = keep
  }

  if (Array.isArray(merged.permissions?.deny)) {
    // Only rules goodvibes installed are retired; a copy the user wrote stays.
    const drop = (p: string) => RETIRED_DENY.includes(p) && wasInstalled(`deny:${p}`)
    for (const p of merged.permissions.deny) if (drop(p)) changes.push(`- permissions.deny: ${p}`)
    merged.permissions.deny = merged.permissions.deny.filter((p: string) => !drop(p))
  }

  for (const list of ['ask', 'deny']) {
    for (const p of tpl.permissions?.[list] ?? []) {
      const have: string[] = merged.permissions?.[list] ?? []
      if (have.includes(p) || wasInstalled(`${list}:${p}`)) continue
      merged.permissions = { ...(merged.permissions ?? {}), [list]: [...have, p] }
      changes.push(`+ permissions.${list}: ${p}`)
    }
  }

  for (const [event, groups] of Object.entries<Json[]>(tpl.hooks ?? {})) {
    for (const g of groups) {
      const id = hookId(g)
      if (!id) continue
      const userGroups: Json[] = merged.hooks?.[event] ?? []
      const idx = userGroups.findIndex(ug => hookId(ug) === id)
      if (idx >= 0) {
        // Only the marked hook is ours; the user's other hooks and fields in that group stay.
        const ug = userGroups[idx]
        const tplHook = g.hooks.find((h: Json) => markerOf(h) === id)
        // The matcher is refreshed only in a group holding nothing but our hook; the user's own hooks keep their routing.
        const ours = 'matcher' in g && ug.hooks.every((h: Json) => markerOf(h) === id)
        const next = { ...ug, ...(ours ? { matcher: g.matcher } : {}), hooks: ug.hooks.map((h: Json) => (markerOf(h) === id ? tplHook : h)) }
        if (!same(ug, next)) {
          userGroups[idx] = next
          changes.push(`~ hooks.${event}: ${id}`)
        }
      } else if (!wasInstalled(`hook:${event}:${id}`)) {
        merged.hooks = { ...(merged.hooks ?? {}), [event]: [...userGroups, g] }
        changes.push(`+ hooks.${event}: ${id}`)
      }
    }
  }
  return { merged, changes }
}

// Keeps previously installed ids so a user's deliberate removal survives later updates.
export async function managedRecord(
  cwd: string,
  templateDir: string,
  prev: Record<string, string[]> = {},
): Promise<Record<string, string[]>> {
  const record: Record<string, string[]> = { ...prev }
  for (const rel of MANAGED_JSON) {
    const tplPath = join(templateDir, rel)
    const destPath = join(cwd, rel)
    if (!existsSync(tplPath) || !existsSync(destPath)) continue
    let content: unknown
    try {
      content = JSON.parse(await readFile(destPath, 'utf-8'))
    } catch {
      continue // unparseable user file: keep the previous record rather than guess
    }
    if (!isJsonObject(content) || shapeError(rel, content)) continue
    const tpl = JSON.parse(await readFile(tplPath, 'utf-8'))
    record[rel] = [...new Set([...(prev[rel] ?? []), ...presentIds(rel, tpl, content)])]
  }
  return record
}
